kobotoolbox/kpi
# Record project history logs for deployment changes

## Checklist

1. [x] If you've added code that should be tested, add tests
2. [ ] If you've changed APIs, update (or create!) the documentation
3. [x] Ensure the tests pass
4. [x] Make sure that your code lints and that you've followed [our coding style](https://github.com/kobotoolbox/kpi/blob/master/CONTRIBUTING.md)
5. [x] Write a title and, if necessary, a description of your work suitable for publishing in our [release notes](https://community.kobotoolbox.org/tag/release-notes)
6. [ ] Mention any related issues in this repository (as #ISSUE) and in other repositories (as kobotoolbox/other#ISSUE)
7. [ ] Open an issue in the [docs](https://github.com/kobotoolbox/docs/issues/new) if there are UI/UX changes

## Description

Record project history log every time a project is deployed, redeployed, archived, or unarchived.

## Notes

Adds a new ProjectHistoryLog proxy subclass of AuditLogs. Creates a log from the `deployment` action of the AssetViewSet, after the serializer saves. 
We create the log in the view rather than the serializer because it's more a request-level concept than an object-level concept. The important part is 'User A made a request to archive/unarchive/deploy/redeploy,' not what happened internally to the models when the user made the request. It's a bit of a violation of "thin view, thick serializer" principle but conceptually I think it makes more sense.
Bulk archive/unarchive will be done elsewhere since it's a different endpoint
Also a small refactoring of test_models to reuse some mocking logic without unused parameters everywhere

To test:
1. Create a new project
2. Deploy the project
3. Open a django shell and run `ProjectHistoryLog.objects.last()`. You should get a log with `metadata.asset_uid` referring to your project's uid and `action=deploy`
4. Redeploy the project
5. A new history log should show up with `action=redeploy`
6. Archive the project
7. A new log should show up with `action=archive`
8. Unarchive the project
9. A new log should show up `action=unarchive`
